### PR TITLE
feat: fast-path physicsspherical to cylindrical

### PR DESCRIPTION
**Pull request by nstarman**
- [ ] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

### CODE REVIEW

**Review by mhvk — COMMENTED** [review]
Makes sense in princple (though perhaps rare?), but if we do this one, then why not also do one for regular spherical representation? And shouldn't we include the reverse ones, from cylindrical to spherical, too?

**Comment by nstarman** [comment]
Including the reverse makes sense to me. I just encountered this in my research and noted it was going through Cartesian, which seemed unnecessary since only cos and sin theta need to be computed to actually do the transform.

**Comment by pllim** [comment]
Why coverage bot says your new code isn't covered?

p.s. Oooo I wonder if #15779 would be useful here.

**Comment by mhvk** [comment]
> Why coverage bot says your new code isn't covered?

It might actually be that we do not explicitly test PhysicsSpherical to Cylindrical...

> p.s. Oooo I wonder if #15779 would be useful here.

Yes, I'm looking forward very much to have benchmarks again!

File: docs/changes/coordinates/16135.feature.rst [hunk 1]
```diff
@@ -0,0 +1,2 @@
+Added a fast path for converting ``PhysicsSphericalRepresentation`` to
```
Comment (astrofrog) [inline docs/changes/coordinates/16135.feature.rst:1]: Are performance improvements normally considered features?

Comment (mhvk) [reply docs/changes/coordinates/16135.feature.rst:1]: Personally don't think a changelog entry is required.

Comment (nstarman) [reply docs/changes/coordinates/16135.feature.rst:1]: Happy to remove. I was on the fence.

Comment (nstarman) [reply docs/changes/coordinates/16135.feature.rst]: Done.

**Review by mhvk — APPROVED** [review]
It would still be nice to do the same for regular `SphericalRepresentation`, but what's here is OK, so in principle it can be done in follow-up.

**Comment by mhvk** [comment]
Note that the failures are real - need to use `allclose` since the two different paths no longer give exactly identical results.

**Comment by nstarman** [comment]
Yikes. CI failure is complete and not due to this PR. I rebased and picked up an upstream issue.

**Comment by neutrinoceros** [comment]
@nstarman this is due to pytest 8.1, see #16146

**Comment by pllim** [comment]
Re: CI -- pytest yanked 8.1. I restarted the jobs and unless you broke it, they should pass now (except devdeps).

File: astropy/coordinates/representation/cylindrical.py [hunk 2]
```diff
@@ -134,6 +134,22 @@ def _scale_operation(self, op, *args):
             result.differentials[key] = differential.__class__(*new_comps, copy=False)
         return result
 
+    def represent_as(self, other_class, differential_class=None):
+        if isinstance(other_class, type):
+            if issubclass(other_class, PhysicsSphericalRepresentation):
+                diffs = self._re_represent_differentials(
+                    other_class, differential_class
+                )
+                r = np.sqrt(self.rho**2 + self.z**2)
```
Comment (eerovaher) [inline astropy/coordinates/representation/cylindrical.py:143]: Wouldn't it be better to use [`np.hypot()`](https://numpy.org/doc/stable/reference/generated/numpy.hypot.html)?

File: astropy/coordinates/tests/test_representation.py [hunk 3]
```diff
@@ -826,6 +826,7 @@ def test_representation_shortcuts(self):
             1 * u.deg, 2 * u.deg, 3 * u.kpc, differentials={"s": difs}
         )
 
+        # Spherical Representation
```
Comment (eerovaher) [inline astropy/coordinates/tests/test_representation.py:829]: I don't find these comments helpful.

File: astropy/coordinates/representation/cylindrical.py (hunk 2 above)
Comment (mhvk) [reply astropy/coordinates/representation/cylindrical.py:143]: `np.hypot` indeed takes proper care when one or the other is small.

File: astropy/coordinates/tests/test_representation.py (hunk 3 above)
Comment (nstarman) [reply astropy/coordinates/tests/test_representation.py]: ✅

File: astropy/coordinates/representation/cylindrical.py (hunk 2 above)
Comment (nstarman) [reply astropy/coordinates/representation/cylindrical.py]: ✅

**Comment by pllim** [comment]
I just applied a label so this runs benchmark implemented in https://github.com/astropy/astropy/pull/15779 (cc @MridulS) assuming this functionality is already benchmarked over at https://github.com/astropy/astropy-benchmarks/tree/main/benchmarks ?

**Comment by pllim** [comment]
Does the benchmark log help at all?

https://github.com/astropy/astropy/actions/runs/8304528883/job/22730265639?pr=16135

**Comment by pllim** [comment]
@MridulS , the log actually says

```
SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
Error: PERFORMANCE DECREASED.
```

but the job is green. Is this expected?

**Comment by MridulS** [comment]
Ahhhh. No it's supposed to give a red cross. I guess in the new iterations I removed that logic. I'll send a PR tomorrow to fix that.

**Comment by mhvk** [comment]
Was looking back at this PR to see its status: @eerovaher - are you happy with @nstarman's changes following your comments?

On the benchmarks, in principle it would be bad if this *decreased* performance, but looking at the log, I see
```
| Change   | Before [d3a39fa1]    | After [ce2f5fbb]    |   Ratio | Benchmark (Parameter)          |
|----------|----------------------|---------------------|---------|--------------------------------|
| +        | 54.7±10μs            | 209±8μs             |    3.81 | units.time_quantity_times_unit |
```
This is indeed a large change but has absolutely nothing to do with this PR. Which in itself is perhaps worrying? (Perhaps best discussed elsewhere, though!)

**Comment by pllim** [comment]
Hmm. I don't see a big jump here:

https://spacetelescope.github.io/bench/astropy-benchmarks/#units.time_quantity_times_unit

**Comment by pllim** [comment]
Might wanna rebase again, just to be sure.

**Comment by mhvk** [comment]
Sounds like it was a fluke. Since the test was in `units`, unrelated to the coordinates stuff here, I think there is no need to re-test. Though we probably need some representation tests, so one can see the benefit of this PR!

**Comment by nstarman** [comment]
Rebased :)

**Comment by eerovaher** [comment]
I'm not seeing any performance increases in the benchmark suite. The most likely explanation is that the `PhysicsSphericalRepresentation` to `CylindricalRepresentation` conversion is not included in the benchmarks, but I don't know what is actually included and because the pull request that implemented benchmarking was merged without any documentation then I can't look it up either.

**Comment by pllim** [comment]
The PR benchmark basically runs https://github.com/astropy/astropy-benchmarks/tree/main/benchmarks but in a *relative* way instead of absolute.

Not sure how much this issue helps to explain but it has links: https://github.com/astropy/astropy-benchmarks/issues/106

**Comment by pllim** [comment]
> not included in the benchmarks

Well, that is a problem... We cannot benchmark something that isn't written.

**Comment by nstarman** [comment]
On my computer I consistently get a 5% speed boost.

**Comment by eerovaher** [comment]
Did you see the 5% boost with the code I posted or did you run something else?

**Comment by nstarman** [comment]
Also, I discovered there's another good reason for this PR.

Pre this PR
```python
import astropy.units as u
import astropy.coordinates as cs

>>> cyl = cs.CylindricalRepresentation(rho=0 * u.kpc, phi=270 * u.deg, z=3 * u.kpc)
>>> cyl.represent_as(cs.PhysicsSphericalRepresentation).phi.deg
180.0
```

With this PR
```python
import astropy.units as u
import astropy.coordinates as cs

>>> cyl = cs.CylindricalRepresentation(rho=0 * u.kpc, phi=270 * u.deg, z=3 * u.kpc)
>>> cyl.represent_as(cs.PhysicsSphericalRepresentation).phi.deg
270.0
```

Consider the transformation (https://en.wikipedia.org/wiki/Cylindrical_coordinate_system)



The `phi` value is conserved through the transformation. Technically at `r=0`, `phi` is not well defined, however I think that keeping `phi` is better than changing it to a value determined by the `Cartesian -> Spherical` conventions.

**Comment by eerovaher** [comment]
I agree with that reasoning. Do add that example to the tests.

**Comment by mhvk** [comment]
On the performance: I think there is a reasonable benefit only when no differentials are involved - those needs a lot more calculations and are not improved by this PR.

**Review by eerovaher — COMMENTED** [review]
At this point I just have a few suggestions regarding the tests.

File: astropy/coordinates/tests/test_representation.py [hunk 4]
```diff
@@ -1380,6 +1399,39 @@ def test_transform(self):
         assert_allclose_quantity(s3.z, expected.z)
         assert_allclose_quantity(s3.rho, expected.rho)
 
+    def test_representation_shortcuts(self):
+        """Test that shortcuts in ``represent_as`` don't fail."""
```
[the remark this answers is not in the document]
Comment (nstarman) [reply astropy/coordinates/tests/test_representation.py:1403]: I copied the name directly from the other representation tests, which are structured the same. To change this name we should refactor all the `test_representation_shortcuts` tests.

[the remark this answers is not in the document]
Comment (eerovaher) [reply astropy/coordinates/tests/test_representation.py:1403]: I think we should change the names, but doing that is a separate task and in this pull request it is indeed better to be consistent with what is already present.

**Review by mhvk — APPROVED** [review]
Looks good!

**Pull request merged**